src/dbnode/storage/index.go

```diff
@@ -1554,52 +1605,107 @@ func (i *nsIndex) queryWithSpan(
+			block:      block,
+		}
+	}
+
+	defer func() {
+		for _, iter := range blockIters {
+			// safe to call Close multiple times, so it's fine to eagerly close in the loop below and here.
+			_ = iter.iterCloser.Close()
+		}
+	}()
+
+	// waitForPermit waits for a permit. returns true if the permit was acquired and the wait time.
+	waitForPermit := func() (bool, time.Duration) {
+		queryCanceled := func() bool {
+			return opts.LimitsExceeded(results.Size(), results.TotalDocsCount()) || state.hasErr()
+		}
+		// make sure the query hasn't been canceled before waiting for a permit.
+		if queryCanceled() {
+			return false, 0
+		}
+		startWait := time.Now()
+		err := permits.Acquire(ctx)
+		waitTime := time.Since(startWait)
+		if err != nil {
+			state.addErr(err)
+			return false, waitTime
+		}
+		// make sure the query hasn't been canceled while waiting for a permit.
+		if queryCanceled() {
+			permits.Release()
+			return false, waitTime
+		}
+		return true, waitTime
+	}
+
+	// We're looping through all the blocks that we need to query and kicking
+	// off parallel queries which are bounded by the permits maximum
+	// concurrency. It's possible at this point that we've completed querying one or more blocks and already exhausted
+	// the maximum number of results that we're allowed to return. If thats the case, there is no value in kicking off
+	// more parallel queries, so we break out of the loop.
+	for _, blockIter := range blockIters {
+		// Capture for async query execution below.
+		blockIter := blockIter
+
+		// acquire a permit before kicking off the goroutine to process the iterator. this limits the number of
+		// concurrent goroutines to # of permits + large queries that needed multiple iterations to finish.
+		acq, waitTime := waitForPermit()
+		blockIter.waitTime += waitTime
+		if !acq {
 			break
 		}
 
-		// Calculate time spent waiting for a worker
 		wg.Add(1)
-		scheduleResult := i.queryWorkersPool.GoWithContext(ctx, func() {
-			startProcessing := time.Now()
-			execBlockFn(ctx, block, query, opts, &state, results, logFields)
-			i.metrics.queryMetrics.blockProcessingTime.RecordDuration(time.Since(startProcessing))
-			wg.Done()
-		})
-		totalWaitTime += scheduleResult.WaitTime
-		if !scheduleResult.Available {
-			state.Lock()
-			state.multiErr = state.multiErr.Add(index.ErrCancelledQuery)
-			state.Unlock()
-			// Did not launch task, need to ensure don't wait for it
+		// kick off a go routine to process the entire iterator.
+		go func() {
+			first := true
+			for !blockIter.iter.Done() {
+				// if this is not the first iteration of the iterator, there were more results than the
+				// MaxResultsPerPermit, so need to acquire another permit.
+				if !first {
+					acq, waitTime := waitForPermit()
+					blockIter.waitTime += waitTime
+					if !acq {
+						break
+					}
+				}
+				first = false
+				startProcessing := time.Now()
+				execBlockFn(ctx, blockIter.block, blockIter.iter, opts, &state, results, logFields)
```

Comment:
Maybe state should just be a pointer type so you don't need to `&` at callsites?